### ISSUE TITLE
GRAPE on Heterogenous graphs

### ISSUE THREAD

**Issue by mvisani** [AnacletoLAB/grape#42]
$\begin{pmatrix}
\end{omatrix}$

**Comment by LucaCappelletti94** [comment]
That's a tad enigmatic as GitHub issues go. Care to expand?

**Comment by mvisani** [comment]
Sorry about that ! I just started the issue and stupidly sent it ! Allow me to update it in a few minutes ;)

**Comment by LucaCappelletti94** [comment]
So for starters:

1. Graphs with multiple edge types and node types and multiple edges between any two given nodes are supported.
2. [Here you can find a tutorial on how to load graphs from CSVs](https://github.com/AnacletoLAB/grape/blob/main/tutorials/Loading_a_Graph_in_Ensmallen.ipynb), though admittedly it does not contain an example with heterogeneous edges - I will add it as soon as I have time. The docstring of the `from csv` method may be already enough, though.
3. That is really something that depends on the embedding method you intend to use and what those features represent. If you already have features for each node, you may not need an embedding in the first place: why do you want an embedding?

Is the task modelled as an edge prediction task? We can do a call next week on the discord channel, just ping me and we can plan it.

Luca

**Comment by mvisani** [comment]
> So for starters:
> 
>     1. Graphs with multiple edge types and node types and multiple edges between any two given nodes are supported.
> 
>     2. [Here you can find a tutorial on how to load graphs from CSVs](https://github.com/AnacletoLAB/grape/blob/main/tutorials/Loading_a_Graph_in_Ensmallen.ipynb), though admittedly it does not contain an example with heterogeneous edges - I will add it as soon as I have time. The docstring of the `from csv` method may be already enough, though.
> 
>     3. That is really something that depends on the embedding method you intend to use and what those features represent. If you already have features for each node, you may not need an embedding in the first place: why do you want an embedding?
> 
> 
> Is the task modelled as an edge prediction task? We can do a call next week on the discord channel, just ping me and we can plan it.
> 
> Luca

Hey ! Thanks for the reply ! 

Yes the task I am trying to achieve is an edge prediction task between a node of type molecule and one of type species (undirected). So far I was able to create a graph with the library. However, if I want to try to use an edge prediction task of your library and I specify `node_features=[embedding_mol, embedding_species]` I get this error : 

```python 
ValueError: The provided node features have 48599 rows but the provided graph Lotus has 71410 nodes. Maybe these features refer to another version of the graph or another graph entirely?
```

Which makes sense since I have 48599 nodes that are of type "molecule" and the rest of type species. I'm thinking the 2 dataframes should be merged but I am wondering if there is an other solution since the number of features is not the same for species of molecules. 

So to answer your question, maybe I don't need embedding but then how do I add features to each node ? 

I'd be glad if we could have a call to have a better understanding of all this. 

Tanks again ! 

Marco

**Comment by LucaCappelletti94** [comment]
I am not sure how you expect a model to ingest such features - could you please describe how you would expect the model of your choosing to work?

**Comment by LucaCappelletti94** [comment]
I am on discord if you'd like to have a call.

**Comment by Filco306** [comment]
I have a similar issue; I am trying to load an undirected graph with edge types *and* weights; my csv that I generate looks like this: 

```bash
head,relation,tail,weight
113091,14,412357,0.7917595
560244,14,1164306,0.7917595
388246,14,1121544,0.7917595
1102500,14,1142585,0.7917595
590896,14,661190,0.7917595
422681,14,501152,0.7917595
754343,14,1105352,0.7917595
639287,14,859151,0.7917595
270949,14,995611,0.7917595
``` 

I use the following snippet, adjusted from one of your tutorials: 

```python3
graph_ = grape.Graph.from_csv(
        # Edges related parameters

        ## The path to the edges list tsv
        edge_path="companykg.csv",
        ## Set the tab as the separator between values
        edge_list_separator=",",
        ## The first rows should NOT be used as the columns names
        edge_list_header=True,
        ## The source nodes are in the first nodes
        sources_column_number=0,
        ## The destination nodes are in the second column
        destinations_column_number=2,
        ## Both source and destinations columns use numeric node_ids instead of node names
        edge_list_numeric_node_ids=True,
        ## The weights are in the third column
        weights_column_number=3,
        edge_type_path="companykg.csv",
        edge_types_column="relation",
        # Nodes related parameters
        ## The path to the nodes list tsv
        # node_path=None,
        ## Set the tab as the separator between values
        # node_list_separator="\t",
        ## The first rows should be used as the columns names
        # node_list_header=True,
        ## The column with the node names is the one with name "node_name".
        # nodes_column="node_name",

        # Graph related parameters
        ## The graph is undirected
        directed=False,
        ## The name of the graph is HomoSapiens
        name="CompanyKG",
        ## Display a progress bar, (this might be in the terminal and not in the notebook)
        verbose=True,
    )
```

However, getting the edge weights seem to not load it. 

```python3
>>> graph_.get_undirected_edge_type_ids()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [27], in <cell line: 1>()
----> 1 graph_.get_undirected_edge_type_ids()

ValueError: The current graph instance does not have edge types.

```

Am I doing something wrong? 

Thank you! :)

**Comment by LucaCappelletti94** [comment]
Weird, I see that I have already added extensive errors for this type of parametrization - which version of Ensmallen are you using? Which OS?

**Comment by Filco306** [comment]
Interesting! 

Version: 
```bash
>>> grape.print_version()
{'GRAPE Version': '0.1.29', 'Python version': '3.10.6', 'Platform': 'Linux-5.4.0-150-generic-x86_64-with-glibc2.31', 'Threads number': 48, 'PyTorch version': '1.13.0', 'PyKEEN version': '1.9.0'}
```

Ensmallen version when I do `pip freeze` is `ensmallen==0.8.36`. I use Ubuntu 20.04.

Should I update the ensmallen package? Or how should I solve it? :)

Thank you for your quick reply?

**Comment by LucaCappelletti94** [comment]
[The latest version on ensmallen is `0.8.65`](https://pypi.org/project/ensmallen/), could you upgrade please?

**Comment by Filco306** [comment]
Ah. Upgraded now, and now I get
```
ValueError: The path to the edge type file (not the edge list!) was provided and is `Some("companykg.csv")`, but you did not provide either `edge_list_edge_types_column` or `edge_list_edge_types_column_number` so to specify which column in the edge list should be loaded. Do note that the file provided to the edge type path should contain the UNIQUE edge types, and not the edge type for each edge. The edge type file is primarily used to ensure all edge types in the edge list are known before starting to process the edge list itself, which allows for additional assumptions and therefore significantly faster processing.
```

**Comment by LucaCappelletti94** [comment]
Ok, which is what I was expecting you to get before. I hope the error I wrote contains sufficient information for you to correct the parametrization, if it remains unclear please do let me know so I can improve the error for the next version of Ensmallen.

**Comment by Filco306** [comment]
Ah, I had to change one parameter, and now it works! However, loading the excellent analysis takes waaaaaay longer; I suppose it takes longer given certain analyses are run that weren't run before? 

Thank you for a great package either way! I love it!

**Comment by LucaCappelletti94** [comment]
As you have now included the edge types, it will incorporate them in the analysis. Most likely, the slowest new step is the [isomorphic edge types detection step, which is similar in nature to the same thing I do for the nodes](https://davidbader.net/publication/2023-cfrb/2023-cfrb.pdf). At some point, I will make a faster version.

**Comment by Filco306** [comment]
Okay! Is there a way to turn off certain analyses such as that one prior to starting the loading of the analysis and leaving in the rest?

**Comment by LucaCappelletti94** [comment]
Hi @Filco306 - I am not sure I have understood your question, could you kindly expand upon it?

**Comment by Filco306** [comment]
Yes absolutely! Is there a way to turn off/skip specifically the isomorphic edge detection step for the analysis summary, so that step is skipped and it finishes faster? That way, I could also test whether it is that step that is the issue.

**Comment by LucaCappelletti94** [comment]
Currently, it runs the complete set of analyses for the whole graph as you have loaded it. In the default analysis that you get when you display the graph object, you cannot provide any parameter, and I tried to make it generally decently fast.

If you are now experiencing long runtimes after having loaded the edge types, it is likely that it is the **isomorphic edge types** analysis that is slow, meaning it is finding many near-isomorphic candidates. Note that this is not the analysis on **isomorphic edges**, which is another thing entirely.

If you **do not load** the edge types, does the analysis complete much faster?

**Comment by Filco306** [comment]
Sorry for my late reply. The analysis still runs fairly slow with the update. I will return with a timing on both!
